src/charm.py

```diff
@@ -431,6 +436,25 @@ def version(self):
             return info.get('version', None)
         return None
 
+    @property
+    def provider_ready(self):
+        """Check status of Prometheus server.
+
+        Status of the Prometheus services is checked by querying
+        Prometheus for its version information. If Prometheus responds
+        with valid information, its status is recorded.
+
+        Returns:
+            True if Prometheus is ready, False otherwise
+        """
+        provided = {'prometheus': self.version}
+        if not self._stored.provider_ready and provided['prometheus']:
+            logger.debug("Prometheus provider is available")
+            logger.debug("Providing : {}".format(provided))
+            self._stored.provider_ready = True
```

Comment:
usually getters do not have side effects

Reply:
If you look at it from that rigid perspective. Or you can look at it as a utility to know if provider is ready or not.

Reply:
Also I plan to explore removing stored state from this charm and have created a [ticket](https://github.com/canonical/prometheus-operator/issues/46) for this.